### PR TITLE
MM-30392 cypress e2e compliance export

### PR DESCRIPTION
**Pull request by furqanmlk**
#### Summary
More Compliance Test Cases have been added.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-30392

### CODE REVIEW

File: e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_multiple_post_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_multiple_post_spec.js:24]:
```suggestion
    it('MM-T1177_01 - Compliance export should include updated posts after editing multiple times, exporting multiple times', () => {
```

File: e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_multiple_post_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_multiple_post_spec.js:47]:
```suggestion
    it('MM-T1177_02 - Compliance export should include updated posts after editing multiple times, exporting multiple times', () => {
```

File: e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_multiple_post_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_multiple_post_spec.js:70]:
```suggestion
    it('MM-T1177_03 - Compliance export should include updated posts after editing multiple times, exporting multiple times', () => {
```

File: e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_multiple_post_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_multiple_post_spec.js:95]:
```suggestion
    it('MM-T1177_04 - Compliance export should include updated posts after editing multiple times, exporting multiple times', () => {
```

File: e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_multiple_post_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_multiple_post_spec.js:123]:
```suggestion
    it('MM-T1177_05 - Compliance export should include updated posts after editing multiple times, exporting multiple times', () => {
```

File: e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_ui_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_ui_spec.js:148]:
```suggestion
    it('MM-T1169_1 - Compliance Export - CSV and Global Relay', () => {
```

File: e2e/cypress/integration/enterprise/system_console/compliance/download_bot_compliance_export_file_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/enterprise/system_console/compliance/download_bot_compliance_export_file_spec.js:60]:
```suggestion
    it('MM-T1175_1 - UserType identifies that the message is posted by a bot', () => {
```

File: e2e/cypress/integration/enterprise/system_console/compliance/download_bot_compliance_export_file_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/enterprise/system_console/compliance/download_bot_compliance_export_file_spec.js:76]:
```suggestion
    it('MM-T1175_2 - UserType identifies that the message is posted by a bot', () => {
```

File: e2e/cypress/support/api_commands.js
Comment (josephbaylon) [inline e2e/cypress/support/api_commands.js:313]: Please move to `e2e/cypress/support/compliance_export_commands.js`

**Review by josephbaylon — CHANGES_REQUESTED** [review]
First pass change requests

File: e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_ui_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_ui_spec.js:172]:
```suggestion
        // # Go to compliance page and disable export
        cy.gotoCompliancePage();
```

File: e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_ui_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_ui_spec.js:181]:
```suggestion
        // # Go to compliance page and enable export
        cy.gotoCompliancePage();
```

File: e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_multiple_post_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_multiple_post_spec.js:37]:
```suggestion
        cy.editPost('edited');
```

**Review by josephbaylon — CHANGES_REQUESTED** [review]
Please update editPost default value

**Comment by furqanmlk** [comment]
> Please update editPost default value

Thanks @josephbaylon 
There is not default value for Edit Post, I just missed one method call :)

File: e2e/cypress/support/compliance_export_commands.js
Comment (josephbaylon) [inline e2e/cypress/support/compliance_export_commands.js:39]:
```suggestion
Cypress.Commands.add('enableComplianceExport', (exportFormat = 'csv') => {
```

File: e2e/cypress/support/compliance_export_commands.js
Comment (josephbaylon) [inline e2e/cypress/support/compliance_export_commands.js:44]:
```suggestion
    cy.findByTestId('exportFormatdropdown').select(exportFormat);
```

File: e2e/cypress/support/compliance_export_commands.js
Comment (josephbaylon) [inline e2e/cypress/support/compliance_export_commands.js:1]: I don't think some of these new commands warrant a place in the global commands. Remember that if you add a global command that looks generic and may be accessed by other tests, this may lead to a lot of confusion since those commands look generic whereas they shouldn't be because they cannot be reused.
 
They are at best helper functions that should be moved to
`enterprise/system_console/compliance/helpers.js`

If you want to retain some of these commands as global functions, they should be under `cypress/support/ui` and following naming convention that relates to compliance export like `uiXXXComplianceExportXXX` and with `d.ts` file included
```
cypress/support/ui/compliance_export.js
cypress/support/ui/compliance_export.d.ts
```

I'll let @saturninoabril chime in on this as he might have a better structure for these commands.

**Review by josephbaylon — CHANGES_REQUESTED** [review]

File: e2e/cypress/support/compliance_export_commands.js
Comment (furqanmlk) [reply e2e/cypress/support/compliance_export_commands.js:1]: Thanks @Joseph
Yeah, I agree some commands are very specific to compliance export tests. 
I initially made local functions but I ended up four spec files and all of them are using these functions. 
I think **helper.js** idea is better. 
Let's wait for Saturn response.

File: e2e/cypress/integration/enterprise/system_console/compliance/download_bot_compliance_export_file_spec.js
Comment (saturninoabril) [inline e2e/cypress/integration/enterprise/system_console/compliance/download_bot_compliance_export_file_spec.js:27]: `pwd` is not used elsewhere and so can declare here.
```suggestion
            const pwd = result.stdout;
```

File: e2e/cypress/integration/enterprise/system_console/compliance/download_bot_compliance_export_file_spec.js
Comment (saturninoabril) [inline e2e/cypress/integration/enterprise/system_console/compliance/download_bot_compliance_export_file_spec.js:80]:
```suggestion
        // # Go to Compliance and enable Run export
```

File: e2e/cypress/integration/enterprise/system_console/compliance/download_bot_compliance_export_file_spec.js
Comment (saturninoabril) [inline e2e/cypress/integration/enterprise/system_console/compliance/download_bot_compliance_export_file_spec.js:107]: These and above two functions were used in other spec. With that, please move to a common `helpers.js` and reuse.

**Review by saturninoabril — CHANGES_REQUESTED** [review]
Thanks @furqanmlk! This is great, especially the way you handled downloaded file.

File: e2e/cypress/support/compliance_export_commands.js
Comment (furqanmlk) [reply e2e/cypress/support/compliance_export_commands.js:162]: > Refer to existing [cy.postBotMessage](https://github.com/mattermost/mattermost-webapp/blob/master/e2e/cypress/support/task_commands.js#L88) and see if it can be of use.

I have moved this method to local as it is taking care of token as well and not being used in other specs.

File: e2e/cypress/support/compliance_export_commands.js
Comment (furqanmlk) [reply e2e/cypress/support/compliance_export_commands.js:15]: I am editing last post using this method

File: e2e/cypress/support/compliance_export_commands.js
Comment (josephbaylon) [inline e2e/cypress/support/compliance_export_commands.js:6]: Please rename, move to `e2e/cypress/support/ui/compliance_export.js` and add corresponding entry to `e2e/cypress/support/ui/compliance_export.d.ts`
```suggestion
Cypress.Commands.add('uiEnableComplianceExport', (exportFormat = 'csv') => {
```

File: e2e/cypress/support/compliance_export_commands.js
Comment (josephbaylon) [inline e2e/cypress/support/compliance_export_commands.js:22]: Please rename, move to `e2e/cypress/support/ui/compliance_export.js` and add corresponding entry to `e2e/cypress/support/ui/compliance_export.d.ts`
```suggestion
Cypress.Commands.add('uiGoToCompliancePage', () => {
```

File: e2e/cypress/support/compliance_export_commands.js
Comment (josephbaylon) [inline e2e/cypress/support/compliance_export_commands.js:27]: Please rename, move to `e2e/cypress/support/ui/compliance_export.js` and add corresponding entry to `e2e/cypress/support/ui/compliance_export.d.ts`
```suggestion
Cypress.Commands.add('uiExportCompliance', () => {
```

File: e2e/cypress/support/compliance_export_commands.js
Comment (josephbaylon) [inline e2e/cypress/support/compliance_export_commands.js:53]: Please rename, move to `e2e/cypress/support/api/common.js` and add corresponding entry to `e2e/cypress/support/api/common.d.ts`
```suggestion
Cypress.Commands.add('apiDownloadFileAndVerifyContentType', (fileURL, contentType = 'application/zip') => {
```

File: e2e/cypress/support/compliance_export_commands.js
Comment (josephbaylon) [inline e2e/cypress/support/compliance_export_commands.js:59]:
```suggestion
        expect(response.headers['content-type']).to.equal(contentType);
```

File: e2e/cypress/support/compliance_export_commands.js
Comment (josephbaylon) [inline e2e/cypress/support/compliance_export_commands.js:11]:
```suggestion
    cy.findByRole('combobox', {name: /export format:/i}).select(exportFormat);
```

File: e2e/cypress/support/index.js
Comment (josephbaylon) [inline e2e/cypress/support/index.js:34]:
```suggestion
```

**Review by josephbaylon — CHANGES_REQUESTED** [review]
Thanks for updating @furqanmlk !! Great work. We're almost there.

Just so we're following the new structure and convention, please rename new commands, move to appropriate files and add corresponding `d.ts` entries for the commands, and update the tests accordingly.

File: e2e/cypress/support/api/common.d.ts
Comment (josephbaylon) [inline e2e/cypress/support/api/common.d.ts:40]:
```suggestion
        apiDownloadFileAndVerifyContentType(fileURL: string, contentType: string): Chainable;
```

File: e2e/cypress/support/ui/compliance_export.d.ts
Comment (josephbaylon) [inline e2e/cypress/support/ui/compliance_export.d.ts:38]:
```suggestion
```

File: e2e/cypress/support/ui/compliance_export.d.ts
Comment (josephbaylon) [inline e2e/cypress/support/ui/compliance_export.d.ts:26]:
```suggestion
        /**
         * Select compliance export format
         * @param {string} exportFormat - compliance export format
         *
         * @example
         *   const EXPORTFORMAT = "Actiance XML";
         *   cy.uiEnableComplianceExport(Compliance Export Format);
         */
```

File: e2e/cypress/support/ui/compliance_export.d.ts
Comment (josephbaylon) [inline e2e/cypress/support/ui/compliance_export.d.ts:32]:
```suggestion
        /**
         * Go to Compliance Page
         */
        uiGoToCompliancePage(): Chainable;
```

File: e2e/cypress/support/ui/compliance_export.d.ts
Comment (josephbaylon) [inline e2e/cypress/support/ui/compliance_export.d.ts:37]:
```suggestion
        /**
         * Click Run Export Compliance and wait for Success status
         */
        uiExportCompliance(): Chainable;
```

File: e2e/cypress/support/ui/index.js
Comment (josephbaylon) [inline e2e/cypress/support/ui/index.js:16]: Please arrange these imports alphabetically. Thanks.

**Review by josephbaylon — APPROVED** [review]
@furqanmlk Just a couple more formatting and rename fixes on `d.t.s` files. I'll go ahead and approve

File: e2e/cypress/support/ui/index.js
Comment (furqanmlk) [reply e2e/cypress/support/ui/index.js]: Thanks @josephbaylon for guidance 👍

File: package-lock.json
Comment (saturninoabril) [inline package-lock.json:13756]: Please revert the changes here since you've not changed anything from `package.json`.

File: e2e/cypress/support/ui/index.js
Comment (saturninoabril) [inline e2e/cypress/support/ui/index.js:16]: Please add end of line.

File: e2e/cypress/support/ui/compliance_export.d.ts
Comment (saturninoabril) [inline e2e/cypress/support/ui/compliance_export.d.ts:17]: Please change below as it was recently changed by https://github.com/mattermost/mattermost-webapp/pull/7512
```suggestion
    interface Chainable {
```

File: e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_ui_spec.js
Comment (saturninoabril) [inline e2e/cypress/integration/enterprise/system_console/compliance/compliance_export_ui_spec.js:173]: I suggests to extract the other step (step 2) into its own new test case in Zephyr. Then this can be as suggested below.

```suggestion
    it('MM-T1169 - Compliance Export - CSV and Global Relay', () => {
```

File: e2e/cypress/integration/enterprise/system_console/compliance/download_compliance_export_file_spec.js
Comment (saturninoabril) [inline e2e/cypress/integration/enterprise/system_console/compliance/download_compliance_export_file_spec.js:227]:
```suggestion
            // # Click post dot menu in center.
            cy.clickPostDotMenu(lastPostId);
```

File: e2e/cypress/support/api/common.js
Comment (saturninoabril) [inline e2e/cypress/support/api/common.js:29]: Per this command implementation, `contentType` may be different. Should this be?
```suggestion
        // * Confirm its content type
```

**Review by saturninoabril — APPROVED** [review]
Thanks @furqanmlk! Left several minor comments but overall looks good to me. I'm approving this now. Feel free to merge once resolved.

**Pull request merged**
